presto-hudi/src/main/java/com/facebook/presto/hudi/split/HudiBackgroundSplitLoader.java

```diff
@@ -0,0 +1,110 @@
+import org.apache.hudi.common.util.HoodieTimer;
+import org.apache.hudi.exception.HoodieException;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+import static com.facebook.presto.hudi.HudiSessionProperties.getSplitGeneratorParallelism;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A runnable to load Hudi splits asynchronously in the background
+ */
+public class HudiBackgroundSplitLoader
+        implements Runnable
+{
+    private static final Logger log = Logger.get(HudiBackgroundSplitLoader.class);
+
+    private final ConnectorSession session;
+    private final ExtendedHiveMetastore metastore;
+    private final HudiTableLayoutHandle layout;
+    private final HoodieTableFileSystemView fsView;
+    private final AsyncQueue<ConnectorSplit> asyncQueue;
+    private final List<String> partitions;
+    private final String latestInstant;
+    private final int splitGeneratorNumThreads;
+    private final ExecutorService splitGeneratorExecutorService;
+
+    public HudiBackgroundSplitLoader(
+            ConnectorSession session,
+            ExtendedHiveMetastore metastore,
+            ExecutorService splitGeneratorExecutorService,
+            HudiTableLayoutHandle layout,
+            HoodieTableFileSystemView fsView,
+            AsyncQueue<ConnectorSplit> asyncQueue,
+            List<String> partitions,
+            String latestInstant)
+    {
+        this.session = requireNonNull(session, "session is null");
+        this.metastore = requireNonNull(metastore, "metastore is null");
+        this.layout = requireNonNull(layout, "layout is null");
+        this.fsView = requireNonNull(fsView, "fsView is null");
+        this.asyncQueue = requireNonNull(asyncQueue, "asyncQueue is null");
+        this.partitions = requireNonNull(partitions, "partitions is null");
+        this.latestInstant = requireNonNull(latestInstant, "latestInstant is null");
+
+        this.splitGeneratorNumThreads = getSplitGeneratorParallelism(session);
+        this.splitGeneratorExecutorService = requireNonNull(splitGeneratorExecutorService, "splitGeneratorExecutorService is null");
+    }
+
+    @Override
+    public void run()
+    {
+        HoodieTimer timer = new HoodieTimer().startTimer();
+        List<HudiPartitionSplitGenerator> splitGeneratorList = new ArrayList<>();
+        List<Future> splitGeneratorFutures = new ArrayList<>();
+        ArrayDeque<String> partitionQueue = new ArrayDeque<>(partitions);
+
+        // Start a number of partition split generators to generate the splits in parallel
+        for (int i = 0; i < splitGeneratorNumThreads; i++) {
+            HudiPartitionSplitGenerator generator = new HudiPartitionSplitGenerator(
+                    session, metastore, layout, fsView, asyncQueue, partitionQueue, latestInstant);
```

Comment:
Is there a need to synchronize the `partitionQueue` or maybe declare it a `ConcurrentLinkedDeque`? All threads will be operating with `partitions` object right.

Reply:
Yes, we don't have to use `AsyncQueue` for fetching partition info and it can be replaced by `ConcurrentLinkedDeque `.  Fixed.